lib/matplotlib/axes/_axes.py

```diff
@@ -212,6 +218,15 @@ def set_xlabel(self, xlabel, fontdict=None, labelpad=None, **kwargs):
         """
         if labelpad is not None:
             self.xaxis.labelpad = labelpad
+        if loc is None:
+            loc = rcParams['xaxis.titlelocation']
+        cbook._check_in_list(('left', 'center', 'right'), loc=loc)
+        if loc == 'right':
+            kwargs.setdefault('x', 1.)
```

Comment:
API: IMHO `loc` should override `x` and `horizontalalignment` parameters. Otherwise, we'd allow mixtures `loc='left', x=1` with probably unwanted behavior (`ha` from `loc` is used, but `x` is overridden`).

Ideally, something like:
~~~
if loc is not None and ('x' in kwargs or 'horizontalalignment' in kwargs):
    raise ValueError("'x' and 'horizontalalignment' cannot be used together with 'loc' in set_xlabel()")
~~~
Might need to check for the abbreviation 'ha' as well.